src/Module/BaseApi.php

```diff
@@ -195,80 +187,108 @@ public static function getRequest(array $defaults)
-				DI::mstdnError()->Forbidden();
-			}
-		}
-
-		if (empty(self::$current_user_id)) {
-			// The execution stops here if no one is logged in
-			api_login(DI::app());
+		$current_token = self::getCachedTokenByBearer();
+		if ($current_token && empty($current_token[$scope])) {
+			Logger::warning('The requested scope is not allowed', ['scope' => $scope, 'application' => self::getCachedTokenByBearer()]);
+			DI::mstdnError()->Forbidden();
 		}
 
-		self::$current_user_id = api_user();
-
-		return (bool)self::$current_user_id;
+		return (bool)api_user();
 	}
 
 	/**
-	 * Get current application
+	 * Returns the current logged user id from the local variable or attempts to retrieve it from the request.
 	 *
-	 * @return array token
+	 * @return int
+	 * @throws HTTPException\ForbiddenException
+	 * @throws HTTPException\InternalServerErrorException
+	 * @throws HTTPException\UnauthorizedException
 	 */
-	protected static function getCurrentApplication()
+	public static function getCachedCurrentUserIdFromRequest(): int
 	{
-		return self::$current_token;
+		static $current_user_id = null;
+
+		if (!is_null($current_user_id)) {
+			return $current_user_id;
+		}
+
+		$current_user_id = self::getCurrentUserIDFromRequest();
+
+		return $current_user_id;
 	}
 
 	/**
 	 * Get current user id, returns 0 if not logged in
 	 *
 	 * @return int User ID
+	 * @throws HTTPException\ForbiddenException
+	 * @throws HTTPException\InternalServerErrorException
+	 * @throws HTTPException\UnauthorizedException
 	 */
-	public static function getCurrentUserID(bool $nologin = false)
+	private static function getCurrentUserIDFromRequest(): int
 	{
-		if (empty(self::$current_user_id)) {
-			self::$current_token = self::getTokenByBearer();
-			if (!empty(self::$current_token['uid'])) {
-				self::$current_user_id = self::$current_token['uid'];
-			} else {
-				self::$current_user_id = 0;
-			}
+		$current_user_id = self::getCurrentUserIDFromToken();
+
+		if (empty($current_user_id)) {
+			// Fetch the user id if logged in - but don't fail if not
+			api_login(DI::app(), false);
+
+			$current_user_id = api_user();
 		}
 
-		if ($nologin) {
-			return (int)self::$current_user_id;
+		return $current_user_id;
+	}
+
+	/**
+	 * Returns the current user id if it is set or returns the user id from the Bearer token.
+	 *
+	 * @return int
+	 */
+	private static function getCurrentUserIDFromToken(): int
+	{
+		$current_token = self::getCachedTokenByBearer();
+		if (!empty($current_token['uid'])) {
+			$current_user_id = $current_token['uid'];
+		} else {
+			$current_user_id = 0;
 		}
 
-		if (empty(self::$current_user_id)) {
-			// Fetch the user id if logged in - but don't fail if not
-			api_login(DI::app(), false);
+		return $current_user_id;
+	}
+
+	/**
+	 * Returns the current authentication token, either from local variable
+	 * or from the Bearer header if it hasn't been retrieved yet
+	 *
+	 * @return array
+	 */
+	protected static function getCachedTokenByBearer(): array
```

Comment:
Why should we cache this value? It isn't used many times during a request.

Reply:
Because it can trigger a database query that should only be performed once during any given API call.

Reply:
But it will be performed only once during any API call, there is no need calling it twice. It would make much more sense caching the fetching of contact data, since this really slows down the output.

Reply:
The token is actually used several times throughout the API, not just for login. For example, the Mastodon `getApplication` method retrieves the token as well, so I believe this small caching is worth it.

Reply:
I still don't like the "cached" in the function name since the uncached function cannot be used from outside this class.

Reply:
The scope of the non-cached method can always be changed if we ever need to use it outside the class, be it in an inherited API module or something else. In the end both methods are doing a similar but different thing and that's what I wanted to emphasize.